### PR TITLE
Draw API

### PR DESCRIPTION
**Pull request by davidpiuva**
An alternative draw API that can be used next to the old API for backward compability.

The only usability problem so far is that you can forget to call `draw` after creating the `DrawState`.

Remaining steps for other pull requests:
* Implement drawing of points, lines and meshes using the Pen cover, source viewport and other new settings.
* Support drawing without GPU.
* Make lots of test cases.
* Slowly get everyone to use the new API while testing everything properly.
* Clean up `pen`, `transform`, `opacity`, `blend`, `viewport`,... from canvas where they do not belong.
* Give matrices in normalized coordinates directly to the vertex shader to reduce the double, tripple, quadruple,... conversions between reference and normalized coordinates.

### CODE REVIEW

**Comment by marcusnaslund** [comment]
Sorry, conflicts due to the removal of `ooc` in all use files.

**Comment by davidpiuva** [comment]
I think this prototype is ready for a first evaluation to get early feedback. @emilwestergren @sebastianbaginski @simonmika

The first commit will be removed before the real PR since the caller always knows the target resolution.

File: source/draw/Map.ooc [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use geometry
+
+Map: abstract class {
```
Comment (davidpiuva) [inline source/draw/Map.ooc]: This is to allow passing shader arguments in GPU independent code. Since no methods are made virtual, it should have no performance impact.

File: source/draw/gpu/opengl/OpenGLCanvas.ooc [hunk 2]
```diff
@@ -28,6 +28,89 @@ OpenGLCanvas: class extends OpenGLSurface {
 	_target: OpenGLPacked
 	_renderTarget: GLFramebufferObject
 	context ::= this _context as OpenGLContext
+	// TODO: Remove once rendering is done in normalized coordinates
```
Comment (davidpiuva) [inline source/draw/gpu/opengl/OpenGLCanvas.ooc]: Converting to reference again in just to decode back to normalized in the vertex shader is messy but first it should start working on top of existing solutions since transform is harder to test.

File: source/draw/gpu/opengl/OpenGLCanvas.ooc [hunk 3]
```diff
@@ -28,6 +28,89 @@ OpenGLCanvas: class extends OpenGLSurface {
...
+		// Set texture
+		if (drawState intputImage)
+			gpuMap add("texture0", drawState intputImage)
+		// Draw
+		gpuMap use()
+		this _bind()
+		this context drawQuad()
+		this _unbind()
+	}
+	draw: override func (action: Func) {
```
Comment (davidpiuva) [inline source/draw/gpu/opengl/OpenGLCanvas.ooc]: The below methods are just moved from another class to clean up. This might become a separate PR.

File: source/draw/gpu/GpuMap.ooc [hunk 4]
```diff
@@ -61,12 +62,8 @@ _Map: class {
 	}
 }
 version(!gpuOff) {
-GpuMap: abstract class {
+GpuMap: abstract class extends Map {
 	_bindings := _Map new()
-	model: FloatTransform3D { get set }
-	view: FloatTransform3D { get set }
-	projection: FloatTransform3D { get set }
-	textureTransform: FloatTransform3D { get set }
```
Comment (davidpiuva) [inline source/draw/gpu/GpuMap.ooc]: Moved to `Map`.

**Comment by davidpiuva** [comment]
@bettinaselig Peer-review usability heuristics?
@emilwestergren Peer-review code quality?

**Comment by bettinaselig** [comment]
What's the difference for the user? Example?

File: source/draw/DrawState.ooc [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+DrawState: cover {
+	target: Image = null
+	intputImage: Image = null
+	map: Map = null
+	opacity := 1.0f
+	_transformNormalized := FloatTransform3D identity
+	viewportLocal := IntBox2D new(0, 0, 0, 0)
+	init: func@ ~default
+	init: func@ ~target (=target)
+	setTarget: func (target: Image) -> This {
```
Comment (emilwestergren) [inline source/draw/DrawState.ooc]: It gets kind of bloated with all these setters and it's just so you can stack them on a single line right?

**Comment by davidpiuva** [comment]
You are one of the users of the API.

File: source/draw/DrawState.ooc [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+use geometry
+import Pen, Image, Map
+
+// Example
+// DrawState new(myImage) setMap(myShader) draw()
+
+DrawState: cover {
+	target: Image = null
+	intputImage: Image = null
```
Comment (emilwestergren) [inline source/draw/DrawState.ooc]: `inputImage`

**Comment by bettinaselig** [comment]
Yes, I know. But what is the difference for me? Could you make an example?

File: source/draw/DrawState.ooc (hunk 5 above)
Comment (davidpiuva) [inline source/draw/DrawState.ooc]: Setters per reference is unpredictable on value types since you might modify a clone and a class would have to be freed.

**Comment by davidpiuva** [comment]
See https://github.com/vidhance/ooc-vidproc/pull/1403/files

In the old API, you change settings in a canvas that is shared between different filters along the graph.
`gpuOriginal uv canvas draw(view, viewportHalf, This _fillUVMap)`

In the new API, you create a `DrawState`, modify settings and call `draw` on the resulting `DrawState`.
`DrawState new(gpuOriginal uv) setViewport(viewportHalf) setMap(This _fillUVMap) draw()`

File: source/draw/DrawState.ooc (hunk 6 above)
Comment (davidpiuva) [inline source/draw/DrawState.ooc]: Updated

File: source/draw/DrawState.ooc [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+		result := this
+		result viewportLocal = viewport
+		result
+	}
+	setInputImage: func (inputImage: Image) -> This {
+		result := this
+		result inputImage = inputImage
+		result
+	}
+	_toNormalized: static func (imageSize: FloatVector2D, transform: FloatTransform3D) -> FloatTransform3D {
```
Comment (davidpiuva) [inline source/draw/DrawState.ooc]: Is there a place for these conversion functions that I have everywhere in my code?

**Comment by davidpiuva** [comment]
Updated with more compact setters for reference transforms. I still have the other setters to compress.

**Comment by bettinaselig** [comment]
Thanks. Ok with me.

**Comment by sebastianbaginski** [comment]
I think you can write setters for covers in more compact way like
```ooc
setInputImage: func (inputImage: Image) -> This {
	this inputImage = inputImage
	this
}
```
If we allow semicolons, then you can even compact it to a single line.

File: source/draw/Map.ooc (hunk 1 above)
Comment (sebastianbaginski) [inline source/draw/Map.ooc]: Do you think you could make a separate PR with this abstract `Map` ?

File: source/draw/gpu/GpuMap.ooc (hunk 4 above)
Comment (sebastianbaginski) [inline source/draw/gpu/GpuMap.ooc]: This is a good change, I think it deserves separate PR.

File: source/draw/Map.ooc (hunk 1 above)
Comment (davidpiuva) [inline source/draw/Map.ooc]: It would take some time to remake commits but I could.

**Comment by davidpiuva** [comment]
I start breaking out the `Map` change.

**Comment by sebastianbaginski** [comment]
OK, thanks

**Comment by davidpiuva** [comment]
@emilwestergren You can continue reviewing.

**Comment by davidpiuva** [comment]
@sebastianbaginski Emil seems to be busy. Can you continue reviewing?

File: source/draw/DrawState.ooc [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+	opacity := 1.0f
+	_transformNormalized := FloatTransform3D identity
+	viewport := IntBox2D new(0, 0, 0, 0)
+	init: func@ ~default
+	init: func@ ~target (=target)
+	setTarget: func (target: Image) -> This {
+		result := this
+		result target = target
+		result
+	}
```
Comment (sebastianbaginski) [inline source/draw/DrawState.ooc]: Maybe compact them to 
```
set* : func( x ) -> This {
    this x = x
    this
}
```
or 
```
set* : func( x ) -> This { this x = x ; this }
```

Comment (davidpiuva) [inline source/draw/DrawState.ooc]: Then we might get a performance problem instead. The current solution is made to work with function inlining and constant propagation.

Comment (davidpiuva) [inline source/draw/DrawState.ooc]: Do you mean that I should just call that directly without using function pointer?

Comment (sebastianbaginski) [inline source/draw/DrawState.ooc]: I mean the `result := this` part is not neccessary, `this` is already passed as a copy because the call is made on a cover (and the function signature does not have a reference indicator `@`).

Comment (davidpiuva) [inline source/draw/DrawState.ooc]: It is against coding convention but okay.

Comment (sebastianbaginski) [inline source/draw/DrawState.ooc]: You mean the semicolon ? Yeah I'm not sure about this, @marcusnaslund ?
Otherwise modifying `this` it is widely used, for example in `Text` cover.

Comment (davidpiuva) [inline source/draw/DrawState.ooc]: Simon said that return values in functions should be named result unless they can be written as a single line expression but that was a year ago.

File: source/draw/DrawState.ooc (hunk 7 above)
Comment (sebastianbaginski) [inline source/draw/DrawState.ooc]: Good question. Static functions in `FloatTransform3D` ?

File: source/draw/DrawState.ooc (hunk 8 above)
Comment (sebastianbaginski) [inline source/draw/DrawState.ooc]: OK. I think he wrote some parts of `Text` cover (or OwnedBuffer) like that himself.

Comment (davidpiuva) [inline source/draw/DrawState.ooc]: Okay then.

File: source/draw/DrawState.ooc [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+	setViewport: func (viewport: IntBox2D) -> This {
+		this viewport = viewport
+		this
+	}
+	setInputImage: func (inputImage: Image) -> This {
+		this inputImage = inputImage
+		this
+	}
+	_toNormalized: static func (imageSize: FloatVector2D, transform: FloatTransform3D) -> FloatTransform3D {
```
Comment (davidpiuva) [inline source/draw/DrawState.ooc]: I tried getting this function to a reusable place around a year ago but it was later removed for not belonging to the draw API. Now I have 5 instances of the same duplicated code in different places.

Maybe a new module for managing coordinate systems using static functions?

Comment (sebastianbaginski) [inline source/draw/DrawState.ooc]: Sure, I guess anything is better than copying the same code in 5 different places

Comment (sebastianbaginski) [inline source/draw/DrawState.ooc]: Will you do that in separate PR ?

Comment (davidpiuva) [inline source/draw/DrawState.ooc]: Yes, I am working on a new PR.

File: source/draw/DrawState.ooc (hunk 8 above)
Comment (marcusnaslund) [inline source/draw/DrawState.ooc]: No semicolons.
The return value may be (and often is) called `result`, but it's not a requirement.

**Comment by davidpiuva** [comment]
I should complete the stub code in GpuCanvas.ooc, squash and rebase on the other pull requests.

**Comment by davidpiuva** [comment]
Rebased

File: source/draw/gpu/opengl/OpenGLCanvas.ooc [hunk 10]
```diff
@@ -28,6 +28,38 @@ OpenGLCanvas: class extends OpenGLSurface {
...
+		this context backend setViewport(viewport)
+		// Set transform
+		gpuMap view = _toLocal * drawState getTransformNormalized() normalizedToReference(this size) * _toLocal
+		// Set model and projection matrices
+		if (this _focalLength > 0.0f) {
+			a := 2.0f * this _focalLength / this size x
+			f := -(this _coordinateTransform e as Float) * 2.0f * this _focalLength / this size y
+			k := (this _farPlane + this _nearPlane) / (this _farPlane - this _nearPlane)
+			o := 2.0f * this _farPlane * this _nearPlane / (this _farPlane - this _nearPlane)
+			gpuMap projection = FloatTransform3D new(a, 0.0f, 0.0f, 0.0f, 0.0f, f, 0.0f, 0.0f, 0.0f, 0.0f, k, -1.0f, 0.0f, 0.0f, o, 0.0f)
```
Comment (sebastianbaginski) [inline source/draw/gpu/opengl/OpenGLCanvas.ooc]: Do you think there is a better place for this math somewhere ? In `FloatTransform3D` maybe ?

Comment (davidpiuva) [inline source/draw/gpu/opengl/OpenGLCanvas.ooc]: I plan to replace all the math that I copied from other draw calls once I have figured out what is a feature and what is a bug.

Comment (sebastianbaginski) [inline source/draw/gpu/opengl/OpenGLCanvas.ooc]: OK :)

File: source/draw/gpu/opengl/OpenGLCanvas.ooc [hunk 11]
```diff
@@ -28,6 +28,38 @@ OpenGLCanvas: class extends OpenGLSurface {
...
+		} else
+			gpuMap projection = FloatTransform3D createScaling(2.0f / this size x, -(this _coordinateTransform e as Float) * 2.0f / this size y, 1.0f)
```
Comment (davidpiuva) [inline source/draw/gpu/opengl/OpenGLCanvas.ooc]: This looks like the place where reference coordinates are converted to normalized coordinates and flipped along Y depending on image format in the orthogonal projection matrix.

File: source/draw/gpu/opengl/OpenGLCanvas.ooc [hunk 12]
```diff
@@ -28,6 +28,38 @@ OpenGLCanvas: class extends OpenGLSurface {
...
+		gpuMap model = this _createModelTransform(IntBox2D new(this size))
```
Comment (davidpiuva) [inline source/draw/gpu/opengl/OpenGLCanvas.ooc]: I don't know what to do with the model transform in normalized coordinates since I don't know if the field of view is vertical or horizontal.

File: source/draw/gpu/opengl/OpenGLCanvas.ooc [hunk 13]
```diff
@@ -28,6 +28,38 @@ OpenGLCanvas: class extends OpenGLSurface {
...
+		viewport := (drawState viewport hasZeroArea) ? IntBox2D new(this size) : drawState viewport
+		this context backend setViewport(viewport)
+		// Set transform
+		gpuMap view = _toLocal * drawState getTransformNormalized() normalizedToReference(this size) * _toLocal
+		// Set model and projection matrices
+		if (this _focalLength > 0.0f) {
+			a := 2.0f * this _focalLength / this size x
+			f := -(this _coordinateTransform e as Float) * 2.0f * this _focalLength / this size y
+			k := (this _farPlane + this _nearPlane) / (this _farPlane - this _nearPlane)
+			o := 2.0f * this _farPlane * this _nearPlane / (this _farPlane - this _nearPlane)
```
Comment (davidpiuva) [inline source/draw/gpu/opengl/OpenGLCanvas.ooc]: I have no idea of who named these variables.

Comment (sebastianbaginski) [inline source/draw/gpu/opengl/OpenGLCanvas.ooc]: Maybe that person wanted to have the same names as in
```ooc
FloatTransform3D: cover {
	a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p: Float
...
```

**Comment by davidpiuva** [comment]
Rebased

File: source/draw/gpu/opengl/OpenGLCanvas.ooc (hunk 11 above)
Comment (emilwestergren) [inline source/draw/gpu/opengl/OpenGLCanvas.ooc]: Correct

File: source/draw/gpu/opengl/OpenGLCanvas.ooc (hunk 12 above)
Comment (emilwestergren) [inline source/draw/gpu/opengl/OpenGLCanvas.ooc]: This is the current model:
**Model matrix** - the world coordinates of the image plane, i.e. the left top corner of the quad is in (-width / 2, height / 2, focalLength)
**View matrix** - the inverse camera matrix, this is where the stabilization transform is applied
**Projection matrix** - The original image plane is reprojected to the new image plane and also the coordinates are translated to normalized coordinates. This should probably be separated into two transforms, one projection and one conversion to normalized coordinates, although you often combine these in OpenGL.

File: source/draw/gpu/opengl/OpenGLCanvas.ooc [hunk 14]
```diff
@@ -28,6 +28,38 @@ OpenGLCanvas: class extends OpenGLSurface {
...
+			gpuMap projection = FloatTransform3D new(a, 0.0f, 0.0f, 0.0f, 0.0f, f, 0.0f, 0.0f, 0.0f, 0.0f, k, -1.0f, 0.0f, 0.0f, o, 0.0f)
+		} else
+			gpuMap projection = FloatTransform3D createScaling(2.0f / this size x, -(this _coordinateTransform e as Float) * 2.0f / this size y, 1.0f)
+		gpuMap model = this _createModelTransform(IntBox2D new(this size))
+		// Set opacity
+		if (drawState opacity < 1.0f)
```
Comment (emilwestergren) [inline source/draw/gpu/opengl/OpenGLCanvas.ooc]: This is extremely ugly and should be fixed in a separate PR.

File: source/draw/gpu/opengl/OpenGLCanvas.ooc (hunk 11 above)
Comment (davidpiuva) [inline source/draw/gpu/opengl/OpenGLCanvas.ooc]: https://github.com/magic-lang/ooc-kean/pull/1264 has been created to clean this up.

File: source/draw/gpu/opengl/OpenGLCanvas.ooc (hunk 13 above)
Comment (davidpiuva) [inline source/draw/gpu/opengl/OpenGLCanvas.ooc]: Double character names (x1, y1, z1, w1, x2, y2, ...) in transforms would make more sense when switching between 3x3 and 4x4 transforms.

Comment (emilwestergren) [inline source/draw/gpu/opengl/OpenGLCanvas.ooc]: @sebastianbaginski is right, it is because of `FloatTransform3D`.

Comment (davidpiuva) [inline source/draw/gpu/opengl/OpenGLCanvas.ooc]: Should we fix `FloatTransform3D` for consistency then? I had to change from 2D to 3D before but luckily I did not use these one dimensional element names.

**Comment by sebastianbaginski** [comment]
If @emilwestergren is happy with this I'll let him set the final review flag.

**Comment by davidpiuva** [comment]
@sebastianbaginski Emil wanted an alternative name for `set` to be used by a new convention when the side effect is returned by return value only.

**Comment by davidpiuva** [comment]
Maybe `pipe` since it looks like piping in unix commands?
`DrawState new(target) pipeMap(shader) draw()`

**Comment by sebastianbaginski** [comment]
I don't know. Maybe `DrawState new(target) updateMap(shader) draw()` ?
`update` is not so short though

**Comment by davidpiuva** [comment]
`give` works together with `get` but marks a difference.

**Comment by sebastianbaginski** [comment]
You mean `DrawState new(target) giveMap(shader) draw()` ?

**Comment by davidpiuva** [comment]
yes

**Comment by sebastianbaginski** [comment]
I think I could live with this or `update`. @emilwestergren ?

**Comment by sebastianbaginski** [comment]
Also @marcusnaslund - any opinions about the name ?

**Comment by davidpiuva** [comment]
Updated with `give`.

**Comment by emilwestergren** [comment]
I think `give` is too confusing. You give the map to the state? Like you pass ownership? The problem here is we want something that makes it more obvious that the modified state is the returned one, not the one you are calling the function on.

**Comment by emilwestergren** [comment]
If we don't have any better suggestions I guess we will go with `set`. However, should we make it a general rule that we don't modify the object for any cover? Any function altering the state of the cover should return a copy.

**Comment by davidpiuva** [comment]
By convention, the address of an argument passed by value should not affect the behaviour since this depends on the current implementation of the compiler. There are cases where the compiler implicitly has to clone the value to a new address causing side effects to be thrown away.

**Comment by davidpiuva** [comment]
Changed back to `set`.

File: source/draw/DrawState.ooc [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+	}
+	setTransformReference: func ~explicit (transform: FloatTransform3D, imageSize: FloatVector2D) -> This {
+		this setTransformNormalized(transform referenceToNormalized(imageSize))
+	}
+	setTransformNormalized: func (transform: FloatTransform3D) -> This {
+		this _transformNormalized = transform
+		this
+	}
+	getTransformNormalized: func -> FloatTransform3D {
+		this _transformNormalized
```
Comment (emilwestergren) [inline source/draw/DrawState.ooc]: If you want to save lines you can put these on a single one.

Comment (davidpiuva) [inline source/draw/DrawState.ooc]: Updated

File: source/draw/gpu/opengl/OpenGLCanvas.ooc [hunk 16]
```diff
@@ -28,6 +28,38 @@ OpenGLCanvas: class extends OpenGLSurface {
 	_target: OpenGLPacked
 	_renderTarget: GLFramebufferObject
 	context ::= this _context as OpenGLContext
+	draw: override func ~DrawState (drawState: DrawState) {
+		// Set shader
```
Comment (marcusnaslund) [inline source/draw/gpu/opengl/OpenGLCanvas.ooc]: If you want these comments, change this one to `//Set shader, viewport, and transform` and remove those on line 34 and 37.

File: source/draw/gpu/opengl/OpenGLCanvas.ooc [hunk 17]
```diff
@@ -28,6 +28,38 @@ OpenGLCanvas: class extends OpenGLSurface {
...
+		if (this _focalLength > 0.0f) {
+			a := 2.0f * this _focalLength / this size x
+			f := -(this _coordinateTransform e as Float) * 2.0f * this _focalLength / this size y
+			k := (this _farPlane + this _nearPlane) / (this _farPlane - this _nearPlane)
+			o := 2.0f * this _farPlane * this _nearPlane / (this _farPlane - this _nearPlane)
+			gpuMap projection = FloatTransform3D new(a, 0.0f, 0.0f, 0.0f, 0.0f, f, 0.0f, 0.0f, 0.0f, 0.0f, k, -1.0f, 0.0f, 0.0f, o, 0.0f)
+		} else
+			gpuMap projection = FloatTransform3D createScaling(2.0f / this size x, -(this _coordinateTransform e as Float) * 2.0f / this size y, 1.0f)
+		gpuMap model = this _createModelTransform(IntBox2D new(this size))
+		// Set opacity
```
Comment (marcusnaslund) [inline source/draw/gpu/opengl/OpenGLCanvas.ooc]: The fact that we're dealing with opacity is obvious.

File: source/draw/gpu/opengl/OpenGLCanvas.ooc [hunk 18]
```diff
@@ -28,6 +28,38 @@ OpenGLCanvas: class extends OpenGLSurface {
...
+		if (drawState opacity < 1.0f)
+			this context backend blend(drawState opacity)
+		else
+			this context backend enableBlend(false)
+		// Set texture
```
Comment (marcusnaslund) [inline source/draw/gpu/opengl/OpenGLCanvas.ooc]: The fact that we're dealing with texture is obvious.

File: source/draw/Canvas.ooc [hunk 19]
```diff
@@ -17,7 +17,7 @@
 use base
 use geometry
 use collections
-import Image, Pen
+import Image, Pen, DrawState
```
Comment (marcusnaslund) [inline source/draw/Canvas.ooc]: Probably not your fault originally, but please put these on three separate lines.

File: source/draw/DrawState.ooc [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use geometry
+import Pen, Image, Map
```
Comment (marcusnaslund) [inline source/draw/DrawState.ooc]: Also here, three lines.

File: source/draw/gpu/opengl/OpenGLCanvas.ooc [hunk 21]
```diff
@@ -28,6 +28,38 @@ OpenGLCanvas: class extends OpenGLSurface {
 	_target: OpenGLPacked
 	_renderTarget: GLFramebufferObject
 	context ::= this _context as OpenGLContext
+	draw: override func ~DrawState (drawState: DrawState) {
+		// Set shader
+		gpuMap: GpuMap = (drawState map == null) ? this context defaultMap : drawState map as GpuMap
```
Comment (marcusnaslund) [inline source/draw/gpu/opengl/OpenGLCanvas.ooc]: `gpuMap: GpuMap = drawState map as GpuMap ?? this context defaultMap`

**Comment by davidpiuva** [comment]
Updated

**Comment by davidpiuva** [comment]
PRInt is done testing.
